### PR TITLE
README updated

### PR DESCRIPTION
**Pull request by javadamiri**
This commit updates and restructures the README file.
The goals are:
* Use the same general structure for mmtk and all bindings.
* Verify the information is correct and up to date

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,44 +1,89 @@
...
+
+## Requirements
+
+We maintain an up to date list of the prerequisite for building MMTk and its bindings in the [mmtk-dev](https://github.com/mmtk/mmtk-dev) repository.
 
 ## Build
-```bash
-cargo +nightly build --no-default-features --features <space seperated features>
+
+Building MMTk needs a nightly Rust toolchain.
```
Comment (steveblackburn) [inline README.md:19]: needs -> requires

File: README.md [hunk 2]
```diff
@@ -1,44 +1,89 @@
...
+As the Rust language and its libraries (crates) are frequently evolving, we recommend using the nightly toolchain specified in the [mmtk-dev](https://github.com/mmtk/mmtk-dev).
+
+```console
+$ # replace nightly-YYYY-MM-DD with the correct toolchain version
```
Comment (steveblackburn) [inline README.md:23]: "the toolchain version specified in mmtk-dev"

File: README.md [hunk 3]
```diff
@@ -1,44 +1,89 @@
...
+$ Export RUSTUP_TOOLCHAIN=nightly-YYYY-MM-DD
+
+$ cargo build --features <space separated features>
 ```
 
-The available features can be seen by examining `Cargo.toml`.
-Currently, there are two different VMs to choose from (JikesRVM and OpenJDK),
-and there are three different plans to choose from (NoGC, SemiSpace and G1).
+It is compulsory to specify one of the available GC plans as a feature at build time. Currently, there are two different plans to choose from:
```
Comment (steveblackburn) [inline README.md:29]: You must specify a GC plan as a feature at build time.

File: README.md [hunk 4]
```diff
@@ -1,44 +1,89 @@
...
+
+* `--features nogc` for NoGC, and
+* `--features semispace` for SemiSpace.
+
+A full list of available features can be seen by examining [`Cargo.toml`](Cargo.toml).
 By passing the `--features` flag to the Rust compiler,
-we conditionally compile VM or plan specific code.
+we conditionally compile plan specific code.
```
Comment (steveblackburn) [inline README.md:36]: plan-specific

File: README.md [hunk 5]
```diff
@@ -1,44 +1,89 @@
...
 You can optionally enable sanity checks by add `sanity` to the set of features
```
Comment (steveblackburn) [inline README.md:37]: adding

File: README.md [hunk 6]
```diff
@@ -1,44 +1,89 @@
...
 You can pass the `--release` flag to the `cargo build` command to use the
 optimizing compiler of Rust for better performance.
 
-Cross compilation can be done by using the `--target` flag.
-
 The artefact produced produced by the build process can be found under
-`target/debug` (or `target/release` for release build).
+`target/debug` (or `target/release` for the release build).
+
+[`ci-build.sh`](.github/scripts/ci-build.sh) shows the builds we are testing in our CI.
```
Comment (steveblackburn) [inline README.md:46]: builds tested by the CI

File: README.md [hunk 7]
```diff
@@ -1,44 +1,89 @@
...
-The API exposed by MMTk can be found under `api/mmtk.h`.
-A client of the memory manager can use MMTk like a C library in the standard way.
-A simple example client that uses MMTk for just allocation can be found under
-`examples/main.c`.
+
+MMTk does not run standalone. You would need to integrate MMTk with a language implementation. You can either try out one of the VM bindings we have been working on, or implement your own binding in your VM for MMTk.
+
+### Try out our current bindings
+
+We are maintaining three VM bindings for MMTk. These bindings are accessible in the following repositories:
```
Comment (steveblackburn) [inline README.md:54]: We maintain

File: README.md [hunk 8]
```diff
@@ -1,44 +1,89 @@
...
+* [JikesRVM](https://github.com/mmtk/mmtk-jikesrvm),
+* [V8](https://github.com/mmtk/mmtk-v8).
+
+For more information on these bindings, please visit their repositories.
+
+### Implement your binding
+
+MMTk provides a bi-directional interface with the language VM.
+
+1. MMTk exposes a set of [API](src/mm/memory_manager.rs). The language VM can call into MMTk by using those APIs.
```
Comment (steveblackburn) [inline README.md:66]: APIs

File: README.md [hunk 9]
```diff
@@ -1,44 +1,89 @@
...
+2. MMTk provides a trait [`VMBinding`](src/vm/mod.rs) that each language VM should implement. MMTk use `VMBinding` to call into the VM.
```
Comment (steveblackburn) [inline README.md:67]: must implement

File: README.md [hunk 10]
```diff
@@ -1,44 +1,89 @@
...
+
+To integrate MMTk with your language implementation, you would need to provide an implementation of `VMBinding`, and
```
Comment (steveblackburn) [inline README.md:69]: you need to

File: README.md [hunk 11]
```diff
@@ -1,44 +1,89 @@
...
-The CI will also build both the 32-bit and 64-bit versions of MMTk for each plan
-to test the `alloc` function.
-
-## VM specific notes
-### JikesRVM
-Please DO NOT build MMTk manually,
-as machine generated code is involved during the build process (e.g. `src/vm/jikesrvm/entrypoint.rs` and `src/vm/jikesrvm/inc.asm`).
-Instead, please invoke the `buildit` script from JikesRVM.
+
+We use both unit tests and VM binding tests to test MMTk in our CI.
```
Comment (steveblackburn) [inline README.md:74]: in the CI

**Review by steveblackburn — APPROVED** [review]
LGTM

I like this.   I only had very minor changes.

**Comment by qinsoon** [comment]
I removed myself from reviewers. If this PR is ready, we can merge it.

**Comment by steveblackburn** [comment]
One more important thing.

Please rename mmtk-dev to mmtk-dev-env.

This difference is very important.    The old name implies that that repo is where all development for mmtk happens.   The new name implies that that repo describes the development environment.

**Pull request merged**